docs/Dialects/Interop/RationaleInterop.md

```diff
@@ -0,0 +1,590 @@
+the `alloc` operation is released. This allows the instance to do some cleanup,
+e.g., when the state type was a pointer and the instance performed some
+`malloc`. Structurally the operation is the same as the `update` operation, but
+without input and output values. The state is also passed by value.
+
+```mlir
+interop.procedural.dealloc cpp %state : !emitc.ptr<!emitc.opaque<"VBar">> {
+^bb0(%arg0: !emitc.ptr<!emitc.opaque<"VBar">>):
+  systemc.cpp.delete %arg0 : !emitc.ptr<!emitc.opaque<"VBar">>
+  interop.return
+}
+```
+
+### Interop Mechanisms
+
+A variety of interop mechanisms can be supported. This allows to perform interop
+on different abstraction levels and only fall back to C-foreign-functions when
+nothing else is supported.
+
+Currently supported interop mechanisms:
+* CFFI: C-foreign-functions, i.e., all interop operations are lowered to C
+  functions and function calls.
+* CPP: textual C++, i.e., the interop operations are lowered to the `systemc`
+  and `emitc` dialects and printed as C++ code. In the future, more dialects
+  (such as `scf`) could be supported.
+
+Adding a new interop mechanism requires changes to the `interop` dialect. At
+a minimum, the table-gen enum as to be modified and a bridging pattern has to
+be added.
+
+### Instance-side Lowering
+
+The instance-side lowering always has to happen before the container-side
+lowering since that pass should create the interop operations that will then
+be picked up by the container-side lowering pass and properly embedded in the
+context of the container operation.
+
+To illustrate how this works, consider a design represented in `hw`, `comb`, and
+`seq`. We want to simulate that design using Verilator and provide a SystemC
+wrapper (basically what Verilator itself can also do). As a first step, the
+top-level module has to be cloned without the region and a
+`systemc.interop.verilated` operation has to be inserted in the body to
+instantiate the previously cloned module (here represented as the extern module
+`@Bar`) as a verilated module. The input and ouput ports get connected 1-1.
+The original design is then exported through `ExportVerilog` and verilated,
+while our wrapper module is lowered by the instance-side lowering.
+
+```mlir
+hw.module.extern @Bar (%a: i32, %b: i32) -> (c: i32)
+
+hw.module @Foo (%x: i32) -> (y: i32) {
+  %c = systemc.interop.verilated "inst" @Bar (a: %x: i32, b: %x: i32) -> (c: i32)
+  hw.output %c : i32
+}
+```
+
+As a result, the above code example is lowered to the following code. This is
+implemented as a pattern on the `systemc.interop.verilated` operation in the
+dialect conversion framework. Note that it is required to provide a rewrite
+pattern for this lowering to enable one-shot lowering of all interop operations
+in a design during a lowering pipeline. The patterns for all instance-side
+interop lowerings of a dialect are provided by a function
+`populate<dialect>InstanceInteropLoweringPatterns(RewritePatternSet&)` exposed
```

Comment:
The PR could use something other than `<dialect>` (to indicate substitution?) to avoid confusion around C++ templates. I may be confused about what this is trying to convey.